taskhawk/backends/gcp.py

```diff
@@ -261,3 +262,17 @@ def requeue_dead_letter(self, num_messages: int = 10, visibility_timeout: int =
                     )
 
             logging.info("Re-queued {} messages".format(len(queue_messages)))
+
+    def health_check(self) -> None:
+        """
+        Checks that subscriber can synchronously pull messages from the queue.
+        """
+        try:
+            self.subscriber.pull(
+                subscription=self._subscription_path,
+                max_messages=1,
```

Comment:
In this particular case wouldn't this one message be lost?

Reply:
For the message to be lost, the subscriber needs to acknowledge it, as the `fetch_and_process_messages` function used by `listen_for_messages` does in https://github.com/cloudchacho/taskhawk-python/blob/master/taskhawk/backends/base.py#L122.

The `delete_message` is a wrapper for `self.publisher.acknowledge` https://github.com/cloudchacho/taskhawk-python/blob/master/taskhawk/backends/gcp.py#L198.

The message will not be removed from the queue otherwise.

Reply:
I've talked with Maru about exposing the `ServiceUnavailable` exception from GCP for health-checking Taskhawk.

Yes, we could do something similar with AWS.

Reply:
Yeah this seems wrong to me. Instead of this, let's keep track of an error count in this class, incrementing it when we get a `ServiceUnavailable` in `pull_messages` and resetting on success.

Reply:
And then simply expose the error count. Apps can then use the error count to implement health checks - for example, a simple implementation would be error > 0 = HC failure.

Reply:
@maroux OK, we can use an error count. But can you elaborate on why you think this solution is wrong? I believe it should work for the intended purpose. Compared with the error count solution, I appreciate it requires additional `subscriber.pull` calls, but it is also a stateless approach.

Reply:
Well, you're affecting real resources just to get health check status. It should ideally be no side effect operation. For example, by pulling a message you're marking it's receive_count up by 1. The next time a real worker receives it, if it checks receive_count to know if this is the first time the message was delivered, you'd get incorrect results.